### PR TITLE
New metric: Normalized Mutual Information Score

### PR DESCRIPTION
**Pull request by matsumotosan**
## What does this PR do?

Fixes part of #2003 

<details>
  <summary>Before submitting</summary>

- [x] Was this **discussed/agreed** via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/torchmetrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to **update the docs**?
- [x] Did you write any new **necessary tests**?

</details>

<details>
  <summary>PR review</summary>

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

</details>

## Did you have fun?

Make sure you had fun coding 🙃

----
:books: Documentation preview :books:: https://torchmetrics--2029.org.readthedocs.build/en/2029/

### CODE REVIEW

File: src/torchmetrics/clustering/normalized_mutual_info_score.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    :math:`M_p(U,V)` is the generalized mean of order :math:`p` of :math:`U` and :math:`V`,
+    and :math:`MI(U,V)` is the mutual information score between clusters :math:`U` and :math:`V`.
+
+    The metric is symmetric, therefore swapping :math:`U` and :math:`V` yields
+    the same mutual information score.
+
+    As input to ``forward`` and ``update`` the metric accepts the following input:
+
+    - ``preds`` (:class:`~torch.Tensor`): either single output float tensor with shape ``(N,)``
+    - ``target`` (:class:`~torch.Tensor`): either single output tensor with shape ``(N,)``
```
Comment (SkafteNicki) [inline src/torchmetrics/clustering/normalized_mutual_info_score.py:43]:
```suggestion
    - ``preds`` (:class:`~torch.Tensor`): single integer tensor with shape ``(N,)`` with predicted cluster labels
    - ``target`` (:class:`~torch.Tensor`): single integer tensor with shape ``(N,)`` with ground truth cluster labels
```

File: src/torchmetrics/clustering/normalized_mutual_info_score.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    is_differentiable = True
+    higher_is_better = None
+    full_state_update: bool = True
+    plot_lower_bound: float = 0.0
+    plot_upper_bound: float = 0.0
+    preds: List[Tensor]
+    target: List[Tensor]
+    contingency: Tensor
+
+    def __init__(self, average_method: Literal["min", "geometric", "arithmetic", "max"], **kwargs: Any) -> None:
```
Comment (SkafteNicki) [inline src/torchmetrics/clustering/normalized_mutual_info_score.py:73]:
```suggestion
    def __init__(self, average_method: Literal["min", "geometric", "arithmetic", "max"] = "arithmetic", **kwargs: Any) -> None:
```
lets pick a default (same as sklearn)

File: src/torchmetrics/functional/clustering/normalized_mutual_info_score.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+# limitations under the License.
+from typing import Literal
+
+import torch
+from torch import Tensor
+
+from torchmetrics.functional.clustering.mutual_info_score import mutual_info_score
+from torchmetrics.functional.clustering.utils import calculate_entropy, calculate_generalized_mean
+
+
```
Comment (SkafteNicki) [inline src/torchmetrics/functional/clustering/normalized_mutual_info_score.py:22]:
```suggestion
def _validate_average_method_arg(average_method: Literal["min", "geometric", "arithmetic", "max"] = "arithmetic") -> None:
     if average_method not in ("min", "geometric", "arithmetic", "max"):
         raise ValueError(f"Expected argument `average_method` to be one of  `min`, `geometric`, `arithmetic`, `max`, but got {average_method}")
```

File: src/torchmetrics/functional/clustering/normalized_mutual_info_score.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    Example:
+        >>> from torchmetrics.functional.clustering import normalized_mutual_info_score
+        >>> target = torch.tensor([0, 3, 2, 2, 1])
+        >>> preds = torch.tensor([1, 3, 2, 0, 1])
+        >>> normalized_mutual_info_score(preds, target, "arithmetic")
+        tensor(0.7919)
+
+    """
+    mutual_info = mutual_info_score(preds, target)
```
Comment (SkafteNicki) [inline src/torchmetrics/functional/clustering/normalized_mutual_info_score.py:44]:
```suggestion
    _validate_average_method_arg(average_method)
    mutual_info = mutual_info_score(preds, target)
```

File: src/torchmetrics/functional/clustering/normalized_mutual_info_score.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+) -> Tensor:
+    """Compute normalized mutual information between two clusterings.
+
+    Args:
+        preds: predicted classes
+        target: ground truth classes
+        average_method: normalizer computation method
+
+    Returns:
+        normalized_mutual_info_score: score between 0.0 and 1.0
```
Comment (SkafteNicki) [inline src/torchmetrics/functional/clustering/normalized_mutual_info_score.py:34]:
```suggestion
        Scalar tensor with normalized mutual info score between 0.0 and 1.0
```

File: src/torchmetrics/functional/clustering/normalized_mutual_info_score.py [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+
+def normalized_mutual_info_score(
+    preds: Tensor, target: Tensor, average_method: Literal["min", "geometric", "arithmetic", "max"] = "arithmetic"
+) -> Tensor:
+    """Compute normalized mutual information between two clusterings.
+
+    Args:
+        preds: predicted classes
+        target: ground truth classes
```
Comment (SkafteNicki) [inline src/torchmetrics/functional/clustering/normalized_mutual_info_score.py:30]:
```suggestion
        preds: predicted cluster labels
        target: ground truth cluster labels
```

File: src/torchmetrics/functional/clustering/normalized_mutual_info_score.py [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+        >>> from torchmetrics.functional.clustering import normalized_mutual_info_score
+        >>> target = torch.tensor([0, 3, 2, 2, 1])
+        >>> preds = torch.tensor([1, 3, 2, 0, 1])
+        >>> normalized_mutual_info_score(preds, target, "arithmetic")
+        tensor(0.7919)
+
+    """
+    mutual_info = mutual_info_score(preds, target)
+    if torch.allclose(mutual_info, torch.tensor(0.0), atol=torch.finfo().eps):
+        return torch.tensor(0.0)
```
Comment (SkafteNicki) [inline src/torchmetrics/functional/clustering/normalized_mutual_info_score.py:46]:
```suggestion
        return mutual_info
```
if this is already zero, no need to create a new tensor

File: src/torchmetrics/functional/clustering/utils.py [hunk 8]
```diff
@@ -11,14 +11,86 @@
...
+
+    Example:
+        >>> from torchmetrics.functional.clustering.utils import calculate_entropy
+        >>> labels = torch.tensor([1, 3, 2, 2, 1])
+        >>> calculate_entropy(labels)
+        tensor(1.0549)
+
+    """
+    if len(x) == 0:
+        return tensor(1.0)
```
Comment (SkafteNicki) [inline src/torchmetrics/functional/clustering/utils.py:41]:
```suggestion
        return tensor(1.0, device=x.device)
```

File: src/torchmetrics/functional/clustering/utils.py [hunk 9]
```diff
@@ -11,14 +11,86 @@
...
+
+    p = torch.bincount(torch.unique(x, return_inverse=True)[1])
+    p = p[p > 0]
+
+    if p.size() == 1:
+        return tensor(0.0)
```
Comment (SkafteNicki) [inline src/torchmetrics/functional/clustering/utils.py:47]:
```suggestion
        return tensor(0.0, device=x.device)
```

File: src/torchmetrics/functional/clustering/utils.py [hunk 10]
```diff
@@ -11,14 +11,86 @@
...
+
+    n = p.sum()
+    return -torch.sum((p / n) * (torch.log(p) - torch.log(n)))
+
+
+def calculate_generalized_mean(x: Tensor, p: Union[int, str]) -> Tensor:
```
Comment (SkafteNicki) [inline src/torchmetrics/functional/clustering/utils.py:53]:
```suggestion
def calculate_generalized_mean(x: Tensor, p: Union[int, Literal["min", "geometric", "arithmetic", "max"]) -> Tensor:
```

File: tests/unittests/clustering/test_utils.py [hunk 11]
```diff
@@ -1,7 +1,7 @@
 # Copyright The Lightning team.
 #
 # Licensed under the Apache License, Version 2.0 (the "License");
-# you may not use this file except in compliance with the License.
+# you may not use this file except in compliancewith the License.
```
Comment (SkafteNicki) [inline tests/unittests/clustering/test_utils.py:4]:
```suggestion
# you may not use this file except in compliance with the License.
```

**Review by SkafteNicki — APPROVED** [review]

File: tests/unittests/clustering/test_normalized_mutual_info_score.py [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+    """Check that metric rejects continuous-valued inputs."""
+    preds, target = _float_inputs
+    with pytest.raises(ValueError, match=r"Expected *"):
+        normalized_mutual_info_score(preds, target, average_method)
+
+
+@pytest.mark.parametrize(
+    ("preds", "target"),
+    [
+        (_single_target_inputs1.preds, _single_target_inputs1.target),
```
Comment (Borda) [inline tests/unittests/clustering/test_normalized_mutual_info_score.py:112]:
```suggestion
        (_single_target_inputs1.preds, _single_target_inputs1.target)
```

**Review by Borda — APPROVED** [review]
